### PR TITLE
Eliminate storage from submitEventProof

### PR DESCRIPTION
**Pull request by sdaveas**

### CODE REVIEW

File: experiment/contractNipopow.sol [hunk 1]
```diff
@@ -345,38 +346,53 @@ contract Crosschain {
...
-                events[hashedBlock].proof,
-                headers,
-                siblings,
-                blockOfInterest
-            )
-        ) {
-            events[hashedBlock].expire = block.number + k;
-            events[hashedBlock].author = msg.sender;
+        // Throws on failure
+        validateInterlink(headers, hashedHeaders, siblings);
```
Comment (gtklocker) [inline experiment/contractNipopow.sol]: this should be a require for clarity, then you also avoid the comment

File: experiment/contractNipopow.sol [hunk 2]
```diff
@@ -345,38 +346,53 @@ contract Crosschain {
...
+        require(
+            hashHeader(headers[headers.length - 1]) == genesisBlockHash,
+            "Proof does not include the genesis block"
+        );
+
+        // Is there any prettier way to do this?
+        bytes32[] memory hashedHeaders = new bytes32[](headers.length);
+        for (uint256 i = 0; i < headers.length; i++) {
+            hashedHeaders[i] = hashHeader(headers[i]);
         }
```
Comment (gtklocker) [inline experiment/contractNipopow.sol]: doesn't look like hashedHeaders are needed in the rest of this function, so can do this outside (maybe inside validateInterlink).

File: experiment/contractNipopow.sol [hunk 3]
```diff
@@ -24,6 +24,7 @@ contract Crosschain {
         address payable author;
         uint256 expire;
         Nipopow proof;
```
Comment (gtklocker) [inline experiment/contractNipopow.sol:26]: proof is still in storage, though unset

**Review by gtklocker — CHANGES_REQUESTED** [review]

File: experiment/contractNipopow.sol (hunk 3 above)
Comment (sdaveas) [reply experiment/contractNipopow.sol]: This it is gone, but the old code of contesting proof still existed at this point, and it needed Nipopow struct

Comment (sdaveas) [reply experiment/contractNipopow.sol:26]: Contesting proof is still at the old version here. It needs the Nipopow structure.

File: experiment/contractNipopow.sol [hunk 4]
```diff
@@ -192,7 +193,7 @@ contract Crosschain {
         bytes32 leaf,
         uint8 mu,
         bytes32[] memory siblings
-    ) internal pure {
+    ) internal pure returns(bool) {
```
Comment (gtklocker) [inline experiment/contractNipopow.sol:196]: formatter didn't run here

**Pull request closed without merging**